segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/ecs/ECSStorage.java

```diff
@@ -0,0 +1,441 @@
+        }
+        return result != null;
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName) {
+        log.info("Creating Segment {}", streamSegmentName);
+
+        if (client.listObjects(config.getEcsBucket(), config.getEcsRoot() + streamSegmentName)
+                  .getObjects().size() != 0) {
+            throw new CompletionException(new StreamSegmentExistsException(streamSegmentName));
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getEcsBucket(),
+                config.getEcsRoot() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getEcsAccessKey(), config.getEcsAccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
+
+        log.info("Created Segment {}", streamSegmentName);
+        return syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentSealedException {
+        log.trace("Writing {} to segment {} at offset {}", length, handle.getSegmentName(), offset);
+
+        if (handle.isReadOnly()) {
+            throw new IllegalArgumentException(handle.getSegmentName());
+        }
+
+        SegmentProperties si = syncGetStreamSegmentInfo(handle.getSegmentName());
+
+        if (si.isSealed()) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        client.putObject(this.config.getEcsBucket(), this.config.getEcsRoot() + handle.getSegmentName(),
+                Range.fromOffsetLength(offset, length), data);
+        return null;
+    }
+
+    private Void syncSeal(SegmentHandle handle) {
+
+        if (handle.isReadOnly()) {
+            log.info("Seal called on a read handle for segment {}", handle.getSegmentName());
+            throw new IllegalArgumentException(handle.getSegmentName());
+        }
+
+        AccessControlList acl = client.getObjectAcl(config.getEcsBucket(),
+                config.getEcsRoot() + handle.getSegmentName());
+        acl.getGrants().clear();
+        acl.addGrants(new Grant[]{new Grant(new CanonicalUser(config.getEcsAccessKey(), config.getEcsAccessKey()),
+                Permission.READ)});
+
+        client.setObjectAcl(
+                new SetObjectAclRequest(config.getEcsBucket(), config.getEcsRoot() + handle.getSegmentName()).withAcl(acl));
+        log.info("Successfully sealed segment {}", handle.getSegmentName());
+        return null;
+    }
+
+    private Void syncConcat(SegmentHandle targetHandle, long offset, String sourceSegment) throws StreamSegmentNotExistsException {
```

Comment:
It looks like this implementation has the same issue we discussed for the file system binding that the client needs to read the data, is that right?